src/librbd/crypto/BlockCrypto.cc

```diff
@@ -0,0 +1,77 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/crypto/BlockCrypto.h"
+#include <alloca.h>
+#include "include/byteorder.h"
+#include "include/ceph_assert.h"
+
+namespace librbd {
+namespace crypto {
+
+template <typename T>
+BlockCrypto<T>::BlockCrypto(DataCryptor<T>* data_cryptor, uint32_t block_size)
+     : m_data_cryptor(data_cryptor), m_block_size(block_size),
+       m_iv_size(data_cryptor->get_iv_size()) {
+  ceph_assert((block_size % data_cryptor->get_block_size()) == 0);
+}
+
+template <typename T>
+void BlockCrypto<T>::crypt(ceph::bufferlist&& data, uint64_t image_offset,
+                        cipher_mode_t mode) {
+  ceph_assert(image_offset % m_block_size == 0);
+  ceph_assert(data.length() % m_block_size == 0);
+
+  unsigned char* iv = (unsigned char*)alloca(m_iv_size);
+  memset(iv, 0, m_iv_size);
+
+  bufferlist src = data;
```

Comment:
Are you trying to use `data` as an in-out parameter? If so, can you use a pointer instead of a rvalue in the parameter list?